src/imgproc/mask.cc

```diff
@@ -0,0 +1,115 @@
+// BoB robotics includes
+#include "common/macros.h"
+#include "imgproc/mask.h"
+#include "imgproc/roll.h"
+
+// Standard C++ includes
+#include <stdexcept>
+
+namespace BoBRobotics {
+namespace ImgProc {
+
+Mask::Mask(cv::Mat mask, const cv::Size &resizeTo)
+{
+    set(std::move(mask), resizeTo);
+}
+
+Mask::Mask(const filesystem::path &imagePath, const cv::Size &resizeTo)
+{
+    set(imagePath, resizeTo);
+}
+
+void
+Mask::apply(const cv::Mat &in, cv::Mat &out) const
+{
+    if (empty()) {
+        out = in;
+    } else {
+        cv::bitwise_and(in, m_Mask, out);
+    }
+}
+
+void
+Mask::combine(const Mask &otherMask, Mask &outputMask) const
+{
+    if (empty()) {
+        outputMask = otherMask;
+    } else if (otherMask.empty()) {
+        outputMask = *this;
+    } else {
+        cv::bitwise_and(m_Mask, otherMask.get(), outputMask.m_Mask);
+    }
+}
+
+size_t
+Mask::countUnmaskedPixels(const cv::Size &size) const
+{
+    return empty() ? size.width * size.height : cv::countNonZero(m_Mask);
+}
+
+bool Mask::empty() const
+{
+    return m_Mask.empty();
+}
+
+const cv::Mat &
+Mask::get() const
+{
+    return m_Mask;
+}
+
+bool
+Mask::isValid(const cv::Size &imageSize) const
+{
+    return empty() || m_Mask.size() == imageSize;
+}
+
+void
+Mask::roll(Mask &out, size_t pixelsLeft) const
+{
+    if (empty()) {
+        out.m_Mask = cv::Mat{};
+    } else {
+        ImgProc::roll(m_Mask, out.m_Mask, pixelsLeft);
+    }
+}
+
+void
+Mask::set(cv::Mat mask, const cv::Size &size)
+{
+    if (mask.empty()) {
+        // Clears mask
+        m_Mask = cv::Mat{};
+        return;
+    }
+
+    // The user has requested a specific size of mask
+    if (!size.empty()) {
+        cv::resize(mask, mask, size);
```

Comment:
I think you might want to explicitly specify the ``cv::INTER_NEAREST`` interpolation mode if we're working with binary images

Reply:
Also, I might _still_ be missing something but if you took a const reference here you could just specify ``m_Mask`` as the target for resize/convert?

Reply:
> I think you might want to explicitly specify the `cv::INTER_NEAREST` interpolation mode if we're working with binary images

Good point!

> Also, I might _still_ be missing something but if you took a const reference here you could just specify `m_Mask` as the target for resize/convert?

As we're ``std::move``ing ``mask`` at the end, effectively it doesn't make any difference, since we're operating on the same chunk of memory. It potentially saves us doing a separate allocation for the output of e.g. ``resize()``.